### PR TITLE
Fix interleaving of ASSERTs with other console output

### PR DESCRIPTION
**Pull request by CraigFe**
... by adding flushes in various places, being sure to flush internal `Format` buffers as well as those of the output channels.

This has the side-effect of ensuring that all `ASSERT` lines are captured in log files (when `--verbose` is not passed), which was previously not the case because of `Format` buffers being flushed after the redirection had stopped.

This supersedes the fix in https://github.com/mirage/alcotest/pull/215, which I now believe is taking the wrong approach: i.e. we should always print `ASSERTS` regardless of test failure, but should also be sure to capture them in the test logs correctly.

### CODE REVIEW

**Comment by samoht** [comment]
For the record, here is the difference between 0.8.3 and this branch on my use-case.

0.8.3:
```
$ dune exec -- test/functoria/test.exe test action 10 -v
Testing functoria.
This run has ID `B9EFFD6C-A823-4C6F-B706-E4316380F6E1`.
[SKIP]              cli              0   read_full_eval.
[SKIP]              cli              1   configure.
[SKIP]              cli              2   describe.
[SKIP]              cli              3   build.
[SKIP]              cli              4   clean.
[SKIP]              cli              5   help.
[SKIP]              cli              6   default.
[SKIP]              cli              7   generated_header.
[SKIP]              package          0   merge.
[SKIP]              package          1   pp.
[SKIP]              graph            0   var_name.
[SKIP]              graph            1   impl_name.
[SKIP]              graph            2   test_graph.
[SKIP]              action           0   sequence.
[SKIP]              action           1   sequence after error.
[SKIP]              action           2   bind passes the correct value to caller code.
[SKIP]              action           3   delete (file exists).
[SKIP]              action           4   delete (file does not exist).
[SKIP]              action           5   mkdir (new dir).
[SKIP]              action           6   mdkir (existing dir).
[SKIP]              action           7   mdkir (existing file).
[SKIP]              action           8   rmdir (non-existing dir).
[SKIP]              action           9   rmdir (existing dir).
 ...                action          10   rmdir (dir with contents).XXX YO
-----------------------------------------------------------------------------------------------------
ASSERT rmdir
-----------------------------------------------------------------------------------------------------
XXX YO
-----------------------------------------------------------------------------------------------------
ASSERT rmdir
-----------------------------------------------------------------------------------------------------
[OK]                action          10   rmdir (dir with contents).
[SKIP]              action          11   with_dir (create file).
[SKIP]              action          12   pwd (root).
[SKIP]              action          13   pwd (vfs).
[SKIP]              action          14   file exists (true).
[SKIP]              action          15   file exists (false).
[SKIP]              action          16   dir exists (true).
[SKIP]              action          17   file exists (true).
[SKIP]              action          18   file exists (false).
[SKIP]              action          19   dir exists (true).
[SKIP]              action          20   size_of (empty).
[SKIP]              action          21   size_of (error).
[SKIP]              action          22   size_of (large).
[SKIP]              action          23   run_cmd fails if the command doesn't exist.
[SKIP]              action          24   run_cmd succeeds if the command exists.
[SKIP]              action          25   write to nonexisting file.
[SKIP]              action          26   write to existing file.
[SKIP]              action          27   write.
Test Successful in 0.001s. 1 test run.
```

this branch:
```
$ dune exec -- test/functoria/test.exe test action 10 -v
Testing functoria.
This run has ID `5411185B-9E23-4EE5-98DE-72A69F78F82C`.
[SKIP]              cli              0   read_full_eval.
[SKIP]              cli              1   configure.
[SKIP]              cli              2   describe.
[SKIP]              cli              3   build.
[SKIP]              cli              4   clean.
[SKIP]              cli              5   help.
[SKIP]              cli              6   default.
[SKIP]              cli              7   generated_header.
[SKIP]              package          0   merge.
[SKIP]              package          1   pp.
[SKIP]              graph            0   var_name.
[SKIP]              graph            1   impl_name.
[SKIP]              graph            2   test_graph.
[SKIP]              action           0   sequence.
[SKIP]              action           1   sequence after error.
[SKIP]              action           2   bind passes the correct value to caller code.
[SKIP]              action           3   delete (file exists).
[SKIP]              action           4   delete (file does not exist).
[SKIP]              action           5   mkdir (new dir).
[SKIP]              action           6   mdkir (existing dir).
[SKIP]              action           7   mdkir (existing file).
[SKIP]              action           8   rmdir (non-existing dir).
[SKIP]              action           9   rmdir (existing dir).
 ...                action          10   rmdir (dir with contents).XXX YO
XXX YO
[OK]                action          10   rmdir (dir with contents).
[SKIP]              action          11   with_dir (create file).
[SKIP]              action          12   pwd (root).
[SKIP]              action          13   pwd (vfs).
[SKIP]              action          14   file exists (true).
[SKIP]              action          15   file exists (false).
[SKIP]              action          16   dir exists (true).
[SKIP]              action          17   file exists (true).
[SKIP]              action          18   file exists (false).
[SKIP]              action          19   dir exists (true).
[SKIP]              action          20   size_of (empty).
[SKIP]              action          21   size_of (error).
[SKIP]              action          22   size_of (large).
[SKIP]              action          23   run_cmd fails if the command doesn't exist.
[SKIP]              action          24   run_cmd succeeds if the command exists.
[SKIP]              action          25   write to nonexisting file.
[SKIP]              action          26   write to existing file.
[SKIP]              action          27   write.
Test Successful in 0.000s. 1 test run.
ASSERT rmdir
ASSERT rmdir
```

To reproduce I think you should try to call `Fmt.epr` in the code being tested and `Alcotest.check`.

**Comment by CraigFe** [comment]
@samoht, @icristescu: this PR has now changed to supersede https://github.com/mirage/alcotest/pull/215 rather than just adding a test. It's ready for review.

File: src/alcotest/core.ml [hunk 1]
```diff
@@ -345,7 +345,9 @@ module Make (M : Monad.S) = struct
     let open Suite in
     let test = suite.fn in
     let pp_event = pp_event t in
+    M.return () >>= fun () ->
     pp_event Fmt.stdout (`Start suite.path);
+    Fmt.(flush stdout) () (* Show event before any test stderr *);
```
Comment (samoht) [inline src/alcotest/core.ml]: why not printing stderr here too?

File: src/alcotest/core.ml [hunk 2]
```diff
@@ -345,7 +345,9 @@ module Make (M : Monad.S) = struct
     let open Suite in
     let test = suite.fn in
     let pp_event = pp_event t in
+    M.return () >>= fun () ->
```
Comment (samoht) [inline src/alcotest/core.ml]: Why is this needed?

**Comment by samoht** [comment]
It's a bit unclear to me where are the points we need flushing : but the current choices seem safe enough.

File: src/alcotest/core.ml (hunk 2 above)
Comment (CraigFe) [reply src/alcotest/core.ml]: This ensures that the printing of the event occurs inside the concurrency monad (which wouldn't be the case otherwise), which is necessary for the Alcotest output to be correctly interleaved with the test output when using `Alcotest_lwt`. It's not tested to be necessary, but it could be.

File: src/alcotest/core.ml (hunk 1 above)
Comment (CraigFe) [reply src/alcotest/core.ml]: There shouldn't be any pending `stderr` here: the `stderr` of each test case is flushed as soon as it finishes and all we've done since then is print events to `stdout`.

**Comment by samoht** [comment]
Travis is reporting useful errors:

```
# File "src/alcotest/alcotest.ml", line 143, characters 13-19:
# Error: This expression has type Format.formatter
#        but an expression was expected of type out_channel
#     ocamlopt src/alcotest/.alcotest.objs/native/alcotest__Core.{cmx,o} (exit 2)
# (cd _build/default && /home/travis/.opam/ocaml-base-compiler.4.04.2/bin/ocamlopt.opt -w -40 -g -I 
```
You probably need a recent version of Fmt (`>=0.8.7`) to have `Fmt.flush`

**Comment by CraigFe** [comment]
> You probably need a recent version of Fmt (`>=0.8.7`) to have `Fmt.flush`

Arf, thanks. Fmt bites me a lot with this sort of thing. Implemented it by hand in https://github.com/mirage/alcotest/pull/228/commits/3544af2d011955c5dd0b2b790752eaa565544a60 to avoid dropping support for OCaml 4.04.

**Comment by CraigFe** [comment]
Not sure what to make of the latest CI error...

```
[ERROR] No switch ocaml-base-compiler.4.08.1 is currently installed. Did you
        mean 'opam switch create ocaml-base-compiler.4.08.1'?
        Installed switches are:

+opam switch create ocaml-base-compiler.4.08.1
```

(https://travis-ci.org/github/mirage/alcotest/jobs/661661199#L378)

File: src/alcotest/core.ml (hunk 2 above)
Comment (icristescu) [reply src/alcotest/core.ml]: Do you mind adding the test which shows that it is necessary? (I don't understand how the alcotest outputs can occur inside the Lwt monad.) There aren't any tests using alcotest_lwt and that expose this bug anyway, it is maybe not a bad idea to add one anyway.

**Comment by samoht** [comment]
I've restarted the failing job.

File: src/alcotest/core.ml (hunk 2 above)
Comment (CraigFe) [reply src/alcotest/core.ml]: Sure, can do :slightly_smiling_face:

I added a test of the previous (broken) state in https://github.com/mirage/alcotest/pull/228/commits/30b47a7aa5a9bfb6a247288a0658895a777d4e21 (with an explanation). Note that the prints are observable but the `assert false` is not: the exception has been caught and hidden behind a bind point, so the failure isn't shown until the computation is run.

I fixed the behaviour in https://github.com/mirage/alcotest/pull/228/commits/847f2271a4cb99bb893189ff030b587b51c8076b. I went a step further and pushed the initial `"Testing ..."` message inside the concurrency monad too, for consistency.

**Comment by CraigFe** [comment]
@samoht: good to go?

**Comment by samoht** [comment]
Thanks!
